### PR TITLE
Changes to support Node Local extents.

### PR DESCRIPTION
**Pull request by hariharan-devarajan**
### Description
Currently, unifyfs clients read node-local data from the server thread which is a bottleneck.
In the case of Laminated files, we can fetch the node-local extents from the server and perform read locally.
With the new proposed changes
- The client has a new configuration for node_local_extents
- On the first read of a laminated file if client.node_local_extents is enabled we fetch extents from the server and update the local extents
- This allows reads to be executed locally
- on server we get our extents using unifyfs_invoke_find_extents_rpc

In these cases, we do only one RPC per request group and fetch extents for all laminated files.

### Motivation and Context
This feature would enable faster reads by clients in the case of laminated files.
Some examples of this are,

- read-only workload where data is cached in unifyfs node-locally.
- producer/consumer use cases like MuMMI are colocated in the same node.

### How Has This Been Tested?
In the API test case, I have enabled node_local_extents to true. This works with existing tests well as for non-laminated files this optimization is not applied. Whereas for laminated files it is.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [X] My code follows the UnifyFS code style requirements.
- [x] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commit messages are properly formatted.

### CODE REVIEW

File: client/src/client_read.c [hunk 1]
```diff
@@ -597,9 +597,67 @@ int process_gfid_reads(unifyfs_client* client,
...
+                cur->next = NULL;
+            }
+            meta->needs_extents_sync = 0;
+        }
+        if (num_request_selected > 0) {
+            /* There are files which are laminated and
+             * require sync of extents */
+            int extent_count = 0;
+            extents_list_t extents = NULL;
+            int ret =
```
Comment (MichaelBrim) [inline client/src/client_read.c:633]: redefinition of 'int ret'

Comment (MichaelBrim) [reply client/src/client_read.c:633]: perhaps use 'int rc' instead. conventionally, 'ret' should be the function return value.

File: common/src/unifyfs_client_rpcs.h [hunk 2]
```diff
@@ -282,6 +283,105 @@ MERCURY_GEN_PROC(unifyfs_heartbeat_in_t,
 MERCURY_GEN_PROC(unifyfs_heartbeat_out_t, ((int32_t)(ret)))
 DECLARE_MARGO_RPC_HANDLER(unifyfs_heartbeat_rpc)
 
+/* unifyfs_node_local_extents_get_rpc (client => server)
+ *
+ * returns node local extents for a file
+ * given a global file id */
+MERCURY_GEN_STRUCT_PROC(unifyfs_index_t,
+                 ((uint32_t)(file_pos))
```
Comment (MichaelBrim) [inline common/src/unifyfs_client_rpcs.h:291]: uint32_t is too small for file/log positions and extent length. use hg_size_t or uint64_t

File: common/src/unifyfs_client_rpcs.h [hunk 3]
```diff
@@ -282,6 +283,105 @@ MERCURY_GEN_PROC(unifyfs_heartbeat_in_t,
...
+                 ((uint32_t)(log_pos))
+                 ((uint32_t)(length))
+                 ((int32_t)(gfid)))
+/* List of extents used in the output of RPC call*/
+typedef struct extents_list {
+    unifyfs_index_t value;
+    struct extents_list* next;
+} *extents_list_t;
```
Comment (MichaelBrim) [inline common/src/unifyfs_client_rpcs.h:299]: hiding the fact that this is a pointer-to-struct in the typedef is confusing for later uses of extents_list_t. I suggest using extents_list_t as the typedef for the structure, and then using explicit pointer logic in the associated code.

File: common/src/unifyfs_client_rpcs.h (hunk 2 above)
Comment (MichaelBrim) [reply common/src/unifyfs_client_rpcs.h:291]: You're not tracking the client_id for the returned extents. The client_id is necessary to know which data log to read. With the current code, I think you have a bug where you are always reading the log for the current client, rather than the log for another client on the same node. There is additional work that will need to be done to setup the logio state for other clients and use it when reading extents from those clients. I think this also means you can't use the existing `extents` seg_tree in the client's file metadata to store the node-local extents, since `seg_tree` has no notion of other clients.

File: t/api/init-fini.c [hunk 4]
```diff
@@ -19,11 +19,14 @@ int api_initialize_test(char* unifyfs_root,
...
     diag("Starting API initialization test");
 
-    int n_configs = 1;
-    unifyfs_cfg_option chk_size = { .opt_name = "logio.chunk_size",
-                                    .opt_value = "32768" };
+    int n_configs = 2;
+    unifyfs_cfg_option options[2];
+    options[0].opt_name = "logio.chunk_size";
+    options[0].opt_value = "32768";
+    options[1].opt_name = "client.node_local_extents";
```
Comment (MichaelBrim) [inline t/api/init-fini.c:26]: I would prefer not to set this option in the API init test.

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
You should make this a draft PR until we can resolve the "reading other client's data" issue I mentioned.

File: client/src/client_read.c [hunk 5]
```diff
@@ -653,15 +653,15 @@ int process_gfid_reads(unifyfs_client* client,
             /* There are files which are laminated and
              * require sync of extents */
             int extent_count = 0;
-            extents_list_t extents = NULL;
-            int ret =
```
Comment (hariharan-devarajan) [inline client/src/client_read.c]: Changed ret to rc as requested

File: t/api/init-fini.c [hunk 6]
```diff
@@ -19,12 +19,10 @@ int api_initialize_test(char* unifyfs_root,
 {
     diag("Starting API initialization test");
 
-    int n_configs = 2;
-    unifyfs_cfg_option options[2];
+    int n_configs = 1;
+    unifyfs_cfg_option options[1];
     options[0].opt_name = "logio.chunk_size";
     options[0].opt_value = "32768";
-    options[1].opt_name = "client.node_local_extents";
```
Comment (hariharan-devarajan) [inline t/api/init-fini.c]: removed from node-local-extents from default cases.

File: common/src/unifyfs_client_rpcs.h [hunk 7]
```diff
@@ -288,26 +288,28 @@ DECLARE_MARGO_RPC_HANDLER(unifyfs_heartbeat_rpc)
...
+                 ((uint64_t)(length))
                  ((int32_t)(gfid))
                  ((int32_t)(log_app_id))
                  ((int32_t)(log_client_id)))
 /* List of extents used in the output of RPC call*/
 typedef struct extents_list {
     unifyfs_client_index_t value;
     struct extents_list* next;
-} *extents_list_t;
+} extents_list_t;
```
Comment (hariharan-devarajan) [inline common/src/unifyfs_client_rpcs.h:301]: removed ptr in typedef and changed relevant code.

File: common/src/unifyfs_client_rpcs.h [hunk 8]
```diff
@@ -288,26 +288,28 @@ DECLARE_MARGO_RPC_HANDLER(unifyfs_heartbeat_rpc)
...
+
+typedef extents_list_t* extents_list_t_ptr;
```
Comment (hariharan-devarajan) [inline common/src/unifyfs_client_rpcs.h:303]: However, the macros don't work with a pointer so typedef is still needed over a pointer. I only use this here and nowhere else.

File: client/src/client_read.c (hunk 1 above)
Comment (hariharan-devarajan) [reply client/src/client_read.c]: I fixed this check my comment below

File: common/src/unifyfs_client_rpcs.h (hunk 2 above)
Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h]: I fixed this check my comment below

File: common/src/unifyfs_client_rpcs.h (hunk 3 above)
Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h]: I fixed this check my comment below

File: t/api/init-fini.c (hunk 4 above)
Comment (hariharan-devarajan) [reply t/api/init-fini.c]: I fixed this check my comment below

File: client/src/client_read.c [hunk 9]
```diff
@@ -597,9 +621,72 @@ int process_gfid_reads(unifyfs_client* client,
...
+                for (int j = 0; j < extent_count; ++j) {
+                    if (fetched_extents_cur->value.log_app_id ==
+                        client->state.app_id) {
+                        int fid = unifyfs_fid_from_gfid(client,
+                                          fetched_extents_cur->value.gfid);
+                        /* get meta for this file id */
+                        unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(
+                                client,
+                                fid);
+                        seg_tree_add(&meta->extents,
```
Comment (MichaelBrim) [inline client/src/client_read.c:674]: probably need to verify `meta != NULL` before adding to `meta->extents`

File: common/src/unifyfs_client_rpcs.h [hunk 10]
```diff
@@ -282,6 +283,109 @@ MERCURY_GEN_PROC(unifyfs_heartbeat_in_t,
...
+                 ((uint64_t)(log_pos))
+                 ((uint64_t)(length))
+                 ((int32_t)(gfid))
+                 ((int32_t)(log_app_id))
+                 ((int32_t)(log_client_id)))
+/* List of extents used in the output of RPC call*/
+typedef struct extents_list {
+    unifyfs_client_index_t value;
+    struct extents_list* next;
+} extents_list_t;
```
Comment (MichaelBrim) [inline common/src/unifyfs_client_rpcs.h:301]: I guess I'm still a little confused by your choice of a list-of-structures here, rather than a simple array-of-structures. Seems like it adds a bunch of extra encoding/decoding overhead versus just transferring the bulk array and using simple indexing.

Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h:301]: So here, I need to send a list of extents on input as well as the output. It was more efficient to transfer it directly as its only metadata.

File: client/src/client_read.c (hunk 9 above)
Comment (hariharan-devarajan) [reply client/src/client_read.c]: ok i added that.

File: client/src/client_read.c [hunk 11]
```diff
@@ -597,9 +621,74 @@ int process_gfid_reads(unifyfs_client* client,
...
     read_req_t* reqs = NULL;
+    if (client->use_node_local_extents) {
+        extents_list_t* list = calloc(1, sizeof(struct extents_list));
+        struct extents_list* cur = list;
+        int num_request_selected = 0;
+        for (int i = 0; i < in_count; ++i) {
+            int fid = unifyfs_fid_from_gfid(client, in_reqs[i].gfid);
+            /* get meta for this file id */
+            unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(client, fid);
+            if (!meta->attrs.is_laminated || !meta->needs_extents_sync) {
```
Comment (MichaelBrim) [inline client/src/client_read.c:632]: need to check `meta != NULL` here too

File: client/src/unifyfs_api_internal.h [hunk 12]
```diff
@@ -37,6 +37,7 @@ typedef struct {
     int pending_unlink;           /* received unlink callback */
 
     int needs_sync;               /* have unsynced writes */
```
Comment (MichaelBrim) [inline client/src/unifyfs_api_internal.h:39]: For clarity, I think we should rename these as:
`needs_sync` -> `needs_write_sync`
`needs_extents_sync` -> `needs_read_sync`

File: common/src/unifyfs_client_rpcs.h (hunk 10 above)
Comment (MichaelBrim) [reply common/src/unifyfs_client_rpcs.h:301]: I think you misunderstood my suggestion. If you just use an array of `unifyfs_client_index_t`, you don't need to write the custom serialization routine and do all the list pointer work. See for example `unifyfs_mread_rpc` where we pass an array of `unifyfs_extent_t` as bulk data.

File: client/src/unifyfs_api_internal.h (hunk 12 above)
Comment (hariharan-devarajan) [reply client/src/unifyfs_api_internal.h:39]: needs_read_sync feels like read data is synchronizing. Maybe we should call it needs_read_metadata/extents_sync

File: common/src/unifyfs_client_rpcs.h (hunk 10 above)
Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h:301]: I can use the bulk data while sending unifyfs_client_index_t but the return doesn't work for me. I am forced to write another RPC from server to client like in case of mread_rpc.

Can you point me to some documentation on how I can return bulk data from an RPC?

File: client/src/client_read.c (hunk 11 above)
Comment (hariharan-devarajan) [reply client/src/client_read.c]: did the change

File: client/src/unifyfs_api_internal.h (hunk 12 above)
Comment (hariharan-devarajan) [reply client/src/unifyfs_api_internal.h]: did the change

File: common/src/unifyfs_client_rpcs.h (hunk 10 above)
Comment (MichaelBrim) [reply common/src/unifyfs_client_rpcs.h:301]: For client-server RPCs, it's all going through shared memory, not using RDMA. My experience is that put/get of whole arrays in shmem is way faster than iterating over a list where you encode/decode each list item.

The size of the array can be dynamic, you only need it to calculate the bulk size (`array size = # elements * sizeof(element)`).

To return bulk data, you just add the bulk args (size and data) to the output struct member list in the `MERCURY_GEN_PROC()`, same as is done for the input struct.

Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h]: Ok, I actually missed pulling data from the request as well as response bulk data that's why it was not working for me.

I fixed that. I had to move the pull function to the common rpc_util file.

I have done the changes and tested it as well

File: client/src/client_read.c [hunk 13]
```diff
@@ -597,9 +621,74 @@ int process_gfid_reads(unifyfs_client* client,
...
+                                         extents[j].file_pos,
+                                         extents[j].file_pos +
+                                         extents[j].length - 1,
+                                         extents[j].log_pos,
+                                         extents[j].log_client_id);
+                        }
+                    }
+                }
+            }
+            free(extents);
```
Comment (MichaelBrim) [inline client/src/client_read.c:686]: should probably move this inside the if block, or check for NULL before calling free()

File: client/src/margo_client.c [hunk 14]
```diff
@@ -888,6 +889,102 @@ int invoke_client_mread_rpc(unifyfs_client* client,
...
+        ret = (int) out.ret;
+        if (ret == (int) UNIFYFS_SUCCESS) {
+            *extent_count = out.extent_count;
+
+            void* out_buffer = pull_margo_bulk_buffer(handle, out.bulk_data,
+                                                  out.bulk_size, NULL);
+            *extents = calloc(out.extent_count, sizeof(unifyfs_client_index_t));
+            unifyfs_client_index_t* out_extents  =
+                    (unifyfs_client_index_t*) out_buffer;
+            for (int i = 0; i < out.extent_count; ++i) {
```
Comment (MichaelBrim) [inline client/src/margo_client.c:961]: why do you need this extra array allocation and copy for loop? just use 
`*extents = (unifyfs_client_index_t*) out_buffer;`

File: common/src/unifyfs_client_rpcs.h [hunk 15]
```diff
@@ -282,6 +283,31 @@ MERCURY_GEN_PROC(unifyfs_heartbeat_in_t,
...
+MERCURY_GEN_STRUCT_PROC(unifyfs_client_index_t,
+                 ((uint64_t)(file_pos))
+                 ((uint64_t)(log_pos))
+                 ((uint64_t)(length))
+                 ((int32_t)(gfid))
+                 ((int32_t)(log_app_id))
+                 ((int32_t)(log_client_id)))
+MERCURY_GEN_PROC(unifyfs_node_local_extents_get_in_t,
+                 ((int32_t)(app_id))
+                         ((int32_t)(client_id))
```
Comment (MichaelBrim) [inline common/src/unifyfs_client_rpcs.h:299]: please fix the alignment for L299-302

File: common/src/unifyfs_client_rpcs.h [hunk 16]
```diff
@@ -282,6 +283,31 @@ MERCURY_GEN_PROC(unifyfs_heartbeat_in_t,
...
+                         ((hg_size_t)(num_req))
+                         ((hg_bulk_t)(bulk_data))
+                         ((hg_size_t)(bulk_size)))
+
+MERCURY_GEN_PROC(unifyfs_node_local_extents_get_out_t,
+                 ((int32_t)(ret))
+                         ((hg_size_t)(extent_count))
```
Comment (MichaelBrim) [inline common/src/unifyfs_client_rpcs.h:306]: please fix the alignment for L306-308

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
@hariharan-devarajan Thanks, this is getting really close to being ready - just a few more requested changes. After that, we'll need to rebase/squash down to a single commit.

File: client/src/margo_client.c (hunk 14 above)
Comment (hariharan-devarajan) [reply client/src/margo_client.c]: Done

File: common/src/unifyfs_client_rpcs.h (hunk 15 above)
Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h]: Done

File: common/src/unifyfs_client_rpcs.h (hunk 16 above)
Comment (hariharan-devarajan) [reply common/src/unifyfs_client_rpcs.h]: Done

**Comment by hariharan-devarajan** [comment]
> @hariharan-devarajan Thanks, this is getting really close to being ready - just a few more requested changes. After that, we'll need to rebase/squash down to a single commit.

I have rebased it :)

File: client/src/margo_client.c [hunk 17]
```diff
@@ -888,6 +889,89 @@ int invoke_client_mread_rpc(unifyfs_client* client,
...
+    hret = margo_get_output(handle, &out);
+    if (hret == HG_SUCCESS) {
+        LOGDBG("Got response ret=%" PRIi32, out.ret);
+        ret = (int) out.ret;
+        if (ret == (int) UNIFYFS_SUCCESS) {
+            *extent_count = out.extent_count;
+            void* out_buffer = pull_margo_bulk_buffer(handle, out.bulk_data,
+                                                  out.bulk_size, NULL);
+            *extents = (unifyfs_client_index_t*) out_buffer;
+
```
Comment (MichaelBrim) [inline client/src/margo_client.c:958]: remove extra line here

File: docs/configuration.rst [hunk 18]
```diff
@@ -63,17 +63,18 @@ a given section and key.
...
+   Key                  Type    Description
+   ==================   ======  =================================================================
+   cwd                  STRING  effective starting current working directory
+   fsync_persist        BOOL    persist data to storage on fsync() (default: on)
+   local_extents        BOOL    service reads from local data if possible (default: off)
+   max_files            INT     maximum number of open files per client process (default: 128)
+   super_magic          BOOL    whether to return UNIFYFS (on) or TMPFS (off) statfs magic (default: on)
+   write_index_size     INT     maximum size (B) of memory buffer for storing write log metadata
+   write_sync           BOOL    sync data to server after every write (default: off)
+   node_local_extents   BOOL    service reads from node local data if possible for laminated files (default: off)
```
Comment (MichaelBrim) [inline docs/configuration.rst:76]: move this up between max_files and super_magic - table is sorted by key

File: server/src/unifyfs_request_manager.c [hunk 19]
```diff
@@ -1440,6 +1444,94 @@ static int process_unlink_rpc(reqmgr_thrd_t* reqmgr,
...
 }
 
+static int process_node_local_extents_get_rpc(reqmgr_thrd_t* reqmgr,
+                                              client_rpc_req_t* req)
+{
+    int ret = UNIFYFS_SUCCESS;
+    unifyfs_node_local_extents_get_in_t* in = req->input;
+    assert(in != NULL);
+    size_t num_req = in->num_req;
+    void* buffer = pull_margo_bulk_buffer(req->handle, in->bulk_data,
```
Comment (MichaelBrim) [inline server/src/unifyfs_request_manager.c:1454]: You should pull the bulk buffer in the RPC handler (`unifyfs_node_local_extents_get_rpc()`) so that the transfer happens in a lightweight argothread. Follow the pattern from `unifyfs_mread_rpc()`, which is do the pull, then on success, assign `client_rpc_req_t.bulk_buf & .bulk_sz`

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
Found a couple more minor things to update.

File: server/src/unifyfs_request_manager.c (hunk 19 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_request_manager.c]: Done

File: docs/configuration.rst (hunk 18 above)
Comment (hariharan-devarajan) [reply docs/configuration.rst]: Done

File: client/src/margo_client.c (hunk 17 above)
Comment (hariharan-devarajan) [reply client/src/margo_client.c]: Done

File: server/src/unifyfs_request_manager.c [hunk 20]
```diff
@@ -1440,6 +1444,90 @@ static int process_unlink_rpc(reqmgr_thrd_t* reqmgr,
...
+    out.ret = (int32_t) ret;
+    out.extent_count = total_chunks;
+    size_t extents_size = total_chunks*sizeof(unifyfs_client_index_t);
+    unifyfs_client_index_t* extents_buffer =
+            malloc(extents_size);
+    extents_list* cur = &out_list;
+    for (int i = 0; i < total_chunks; ++i) {
+        extents_buffer[i].file_pos = cur->value.file_pos;
+        extents_buffer[i].length = cur->value.length;
+        extents_buffer[i].gfid = cur->value.gfid;
```
Comment (MichaelBrim) [inline server/src/unifyfs_request_manager.c:1505]: I think you are missing assignments to the log-related fields of the `unifyfs_client_index_t` here

Comment (hariharan-devarajan) [reply server/src/unifyfs_request_manager.c:1505]: Done

File: client/src/unifyfs_api.c [hunk 21]
```diff
@@ -122,7 +131,17 @@ unifyfs_rc unifyfs_initialize(const char* mountpoint,
             client->use_local_extents = (bool)b;
         }
     }
-
+    /* Determine if we should track all extents on the node-local clients
```
Comment (MichaelBrim) [inline client/src/unifyfs_api.c:134]: Looks like this code block is a duplicate of the one above.

File: server/src/unifyfs_client_rpc.c [hunk 22]
```diff
@@ -911,7 +911,7 @@ static void unifyfs_get_gfids_rpc(hg_handle_t handle)
         /* don't need the input struct any more, so free it now. */
         free(in);
         in = NULL;
-    }
+        }
```
Comment (MichaelBrim) [inline server/src/unifyfs_client_rpc.c:914]: fix spacing here

File: server/src/unifyfs_group_rpc.c [hunk 23]
```diff
@@ -148,7 +149,6 @@ static int get_child_response(coll_request* coll_req,
 
             margo_free_output(chdl, out);
         }
-        free(out);
```
Comment (MichaelBrim) [inline server/src/unifyfs_group_rpc.c:151]: Why was this removed?

File: server/src/unifyfs_group_rpc.c [hunk 24]
```diff
@@ -387,6 +367,8 @@ static int collective_forward(coll_request* coll_req)
     return ret;
 }
 
+
```
Comment (MichaelBrim) [inline server/src/unifyfs_group_rpc.c:370]: remove these empty lines

File: server/src/unifyfs_group_rpc.c [hunk 25]
```diff
@@ -433,8 +415,7 @@ void collective_set_local_retval(coll_request* coll_req, int val)
     }
 }
 
-/* finish collective process by waiting for any child responses and
- * sending parent response (if applicable) */
+/* Forward the collective request to any children */
```
Comment (MichaelBrim) [inline server/src/unifyfs_group_rpc.c:418]: Revert to the previous comment

File: server/src/unifyfs_group_rpc.c [hunk 26]
```diff
@@ -479,9 +462,10 @@ int invoke_bcast_progress_rpc(coll_request* coll_req)
     }
 
     /* get handle to local rpc function */
+    hg_handle_t handle;
```
Comment (MichaelBrim) [inline server/src/unifyfs_group_rpc.c:465]: Why did this change to a local variable? Revert to using the handle saved in the the coll_request

File: server/src/unifyfs_group_rpc.c [hunk 27]
```diff
@@ -491,8 +475,7 @@ int invoke_bcast_progress_rpc(coll_request* coll_req)
          * by a ULT */
         bcast_progress_in_t in;
         in.coll_req = (hg_ptr_t) coll_req;
-        hret = margo_iforward(coll_req->progress_hdl, &in,
-                              &(coll_req->progress_req));
+        hret = margo_forward(handle, &in);
```
Comment (MichaelBrim) [inline server/src/unifyfs_group_rpc.c:478]: Revert to using margo_iforward()

File: server/src/unifyfs_request_manager.c [hunk 28]
```diff
@@ -84,6 +86,8 @@ do { \
  * to process the read replies. It iterates with the client until
  * all incoming read replies have been transferred. */
 
+typedef struct extents_list extents_list;
```
Comment (MichaelBrim) [inline server/src/unifyfs_request_manager.c:89]: this typedef should be in the same place as the struct definition

File: t/api/init-fini.c [hunk 29]
```diff
@@ -20,10 +20,11 @@ int api_initialize_test(char* unifyfs_root,
     diag("Starting API initialization test");
 
     int n_configs = 1;
-    unifyfs_cfg_option chk_size = { .opt_name = "logio.chunk_size",
-                                    .opt_value = "32768" };
+    unifyfs_cfg_option options[1];
```
Comment (MichaelBrim) [inline t/api/init-fini.c:23]: revert to prior code

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
I think most of the requested changes are due to unintended rebase effects.

File: server/src/unifyfs_client_rpc.c (hunk 22 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_client_rpc.c:914]: > fix spacing here

I Checked the bracket matches the else indent.

File: t/api/init-fini.c (hunk 29 above)
Comment (hariharan-devarajan) [reply t/api/init-fini.c]: fixed

File: server/src/unifyfs_request_manager.c (hunk 28 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_request_manager.c]: fixed

File: server/src/unifyfs_group_rpc.c (hunk 27 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_group_rpc.c]: fixed

File: server/src/unifyfs_group_rpc.c (hunk 26 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_group_rpc.c]: fixed

File: server/src/unifyfs_group_rpc.c (hunk 24 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_group_rpc.c]: fixed

File: server/src/unifyfs_group_rpc.c (hunk 23 above)
Comment (hariharan-devarajan) [reply server/src/unifyfs_group_rpc.c]: fixed

File: client/src/unifyfs_api.c (hunk 21 above)
Comment (hariharan-devarajan) [reply client/src/unifyfs_api.c:124]: fixed

Comment (MichaelBrim) [reply client/src/unifyfs_api.c:124]: the below block is still a duplicate. the above block should be reverted to update client->use_local_extents

File: server/src/unifyfs_group_rpc.c [hunk 30]
```diff
@@ -54,7 +55,7 @@ static int forward_child_request(void* input_ptr,
     int ret = UNIFYFS_SUCCESS;
 
     /* call rpc function */
-    double timeout_ms = margo_server_server_timeout_msec;
+    double timeout_ms = UNIFYFS_MARGO_SERVER_SERVER_TIMEOUT_MSEC;
```
Comment (MichaelBrim) [inline server/src/unifyfs_group_rpc.c:58]: You need to revert this whole file to the current dev version and then add your new `#include unifyfs_rpc_util.h`

File: t/api/init-fini.c [hunk 31]
```diff
@@ -21,7 +21,7 @@ int api_initialize_test(char* unifyfs_root,
 
     int n_configs = 1;
     unifyfs_cfg_option chk_size = { .opt_name = "logio.chunk_size",
-                                    .opt_value = "32768" };
+            .opt_value = "32768" };
```
Comment (MichaelBrim) [inline t/api/init-fini.c:24]: revert this file to the dev version

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
There's still a few places where the rebase went wrong. See prior comments

File: client/src/unifyfs_api.c (hunk 21 above)
Comment (MichaelBrim) [reply client/src/unifyfs_api.c:114]: now there is no duplication, but the original block that updates `client->use_local_extents` is missing

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
Just the one remaining issue. Almost there :)

File: client/src/unifyfs_api.c (hunk 21 above)
Comment (hariharan-devarajan) [reply client/src/unifyfs_api.c:126]: ok added that back.

**Review by MichaelBrim — APPROVED** [review]
All looks good now. Thanks @hariharan-devarajan

**Review by adammoody — APPROVED** [review]
